pkg/tools/cli/loadtest/service/search.go

```diff
@@ -17,26 +17,61 @@ package service
 
 import (
 	"context"
+	"sync"
 
+	"github.com/vdaas/vald/apis/grpc/agent/core"
 	"github.com/vdaas/vald/apis/grpc/gateway/vald"
 	"github.com/vdaas/vald/apis/grpc/payload"
+	"github.com/vdaas/vald/internal/errors"
 	"github.com/vdaas/vald/internal/net/grpc"
 	"github.com/vdaas/vald/pkg/tools/cli/loadtest/assets"
+	"github.com/vdaas/vald/pkg/tools/cli/loadtest/config"
 )
 
-func newSearch() (requestFunc, loaderFunc) {
-	return func(dataset assets.Dataset) ([]interface{}, error) {
-			vectors := dataset.Query()
-			requests := make([]interface{}, len(vectors))
-			for j, v := range vectors {
-				requests[j] = &payload.Search_Request{
-					Vector: v,
-				}
+func searchRequestProvider(dataset assets.Dataset) (func() interface{}, int, error) {
+	v := dataset.Query()
+	size := len(v)
+	i := 0
+	m := sync.Mutex{}
+	return func() (ret interface{}) {
+		m.Lock()
+		defer m.Unlock()
```

Comment:
I think atomic operations have less impact on performance than Lock.:thinking:

Reply:
I think that I cannot use atomic package in this case, because this whole operation block must be atomic.
In detail, **i** must have same value in this block, so I think that atomic package cannot apply.